### PR TITLE
Feature/aethel detail view

### PR DESCRIPTION
**Pull request by XanderVertegaal**
A first version of the Sample detail view.

I'm not very proud of how the page is styled at the moment, but at least it is (should be) functional, so we can show something to Kokos and Michaël.

### CODE REVIEW

File: backend/aethel_db/views.py [hunk 1]
```diff
@@ -112,6 +114,74 @@ def item_contains_query_string(item: LexicalItem) -> bool:
         return response_object.json_response()
 
 
+class AethelDetailError(Enum):
+    NO_QUERY_INPUT = "NO_QUERY_INPUT"
```
Comment (XanderVertegaal) [inline backend/aethel_db/views.py:118]: None of these errors are actually expected, but it is good to have them, I think.

File: backend/spindle/utils.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+from aethel.frontend import LexicalPhrase
+
+
+def serialize_phrases_with_infix_notation(
```
Comment (XanderVertegaal) [inline backend/spindle/utils.py:4]: I needed this for my Sample page as well, so I've extracted it to a separate file. `utils.py` is not the most helpful filename, but this function _is_ a helpful utility function.

File: frontend/src/app/menu/menu.component.spec.ts [hunk 3]
```diff
@@ -1,18 +1,22 @@
...
+import { RouterModule } from "@angular/router";
+import { routes } from "../routes";
+describe("MenuComponent", () => {
     let component: MenuComponent;
     let fixture: ComponentFixture<MenuComponent>;
 
     beforeEach(waitForAsync(() => {
         TestBed.configureTestingModule({
             declarations: [MenuComponent],
-            imports: [NoopAnimationsModule, RouterTestingModule, HttpClientTestingModule]
```
Comment (XanderVertegaal) [inline frontend/src/app/menu/menu.component.spec.ts:15]: Apparently, the RouterTestingModule is deprecated. The regular RouterModule should take care of everything.

File: frontend/src/app/aethel/aethel.component.ts [hunk 4]
```diff
@@ -45,21 +49,41 @@ export class AethelComponent implements OnInit {
                 }
                 this.rows = this.addUniqueKeys(response.results);
             });
+
+        // If the route has a query parameter, we run a search with that query immediately.
+        this.route.queryParams
```
Comment (bbonf) [inline frontend/src/app/aethel/aethel.component.ts:54]: I don't mean to keep being difficult about `rxjs`, but in this case the following seems much simpler to me:

```
let params = new URLSearchParams(location.search);
if (params.has('query')) {
    this.form.controls.aethelInput.setValue(params.get('query'));
    this.search();
}
```

File: frontend/src/app/sample/sample.component.html [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    </div>
+    <div class="message-body" [innerHTML]="sample.term | proof"></div>
+</section>
+
+<table class="table is-fullwidth">
+    <thead>
+        <tr>
+            <th i18n>#</th>
+            <th i18n>Phrase</th>
+            <th i18n>Proof</th>
```
Comment (bbonf) [inline frontend/src/app/sample/sample.component.html:21]: I think this column should be renamed to "term"

File: frontend/src/app/sample/sample.component.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+    constructor(
+        private route: ActivatedRoute,
+        private apiService: AethelApiService,
+        private router: Router,
+    ) {}
+
+    public routeToAethel(items: LexicalPhrase["items"]): void {
+        const combined = items.map((item) => item.word).join(" ");
+        this.router.navigate(["/æthel"], {
```
Comment (bbonf) [inline frontend/src/app/sample/sample.component.ts:28]: using æ in the url makes me uneasy, but I can't fully explain why

File: backend/aethel_db/views.py [hunk 7]
```diff
@@ -112,6 +114,74 @@ def item_contains_query_string(item: LexicalItem) -> bool:
...
+            response = AethelDetailResponse(error=AethelDetailError.NO_QUERY_INPUT)
+            return response.json_response()
+
+        samples = dataset.find_by_name(query_input)
+
+        if len(samples) == 0:
+            response = AethelDetailResponse(error=AethelDetailError.SAMPLE_NOT_FOUND)
+            return response.json_response()
+
+        if len(samples) > 1:
```
Comment (bbonf) [inline backend/aethel_db/views.py:178]: I'd say it's the responsibility of the data layer (and not the query interface) to make sure names are unique

File: backend/aethel_db/views.py [hunk 8]
```diff
@@ -112,6 +114,74 @@ def item_contains_query_string(item: LexicalItem) -> bool:
         return response_object.json_response()
 
 
+class AethelDetailError(Enum):
+    NO_QUERY_INPUT = "NO_QUERY_INPUT"
+    SAMPLE_NOT_FOUND = "SAMPLE_NOT_FOUND"
+    MULTIPLE_FOUND = "MULTIPLE_FOUND"
+
+
+aethelDetailStatusCodes = {
```
[the remark this answers is not in the document]
Comment (XanderVertegaal) [reply backend/aethel_db/views.py:123]: This bit got shifted around to/from the frontend; sorry!

File: backend/aethel_db/views.py (hunk 7 above)
Comment (XanderVertegaal) [reply backend/aethel_db/views.py:178]: I agree, but can we rely on this?

Would you rather just use `samples = dataset.find_by_name(query_input)[0]` and assume we will ever only get one result?

File: frontend/src/app/sample/sample.component.ts (hunk 6 above)
Comment (XanderVertegaal) [reply frontend/src/app/sample/sample.component.ts:28]: Me too, and hearing that you find it strange as well, I'll make sure to change it.

(Fun fact: using æ in the template confuses the Angular language service, apparently.)

File: frontend/src/app/aethel/aethel.component.ts (hunk 4 above)
Comment (XanderVertegaal) [reply frontend/src/app/aethel/aethel.component.ts:54]: You're right! Nice one :D

Comment (XanderVertegaal) [reply frontend/src/app/aethel/aethel.component.ts:64]: Actually, with the dynamic queryParams that trigger searches, we do need a reactive approach, so I have kept the observable in place for now.

**Comment by XanderVertegaal** [comment]
In response to your general comments:

>     * Either add a button for going back to search results from detail view (or use the browser's back button), or open detail view in a new tab.

Back button has been added; with the query parameters being reactive, the user should also just be able to press the 'Back' button in their browser.

>     * I think it might not be clear from the search results page that results are grouped based on word-lemma-type tuples. Not sure how to best solve that.

This is something that we could include with the text (that we have to add anyway) that explains what this dataset is.

>     * "Search in Aethel" button should maybe be aligned to the right

Done!
 
>     * "Search in Aethel" button doesn't work when result token is less than 3 letters long (easy to test with punctuation)

Good find. I now hide the items that are shorter than 3 letters. I played around with adding some text instead that said 'Too short to search' or something, but that looked a bit messy. We should probably indicate why not every phrase has a 'Search in Æthel button now'.

>     * Aethel query parameter doesn't update on 

Now it does! I ended up using `router.navigate` as you suggested! Turns out that that does not refresh the page if you stay on the same path.

**Review by bbonf — APPROVED** [review]

**Pull request merged**